src/translate.py

```diff
@@ -185,9 +185,56 @@ def as_function_ptr(expr: "Expression") -> "Expression":
     return as_type(expr, Type.ptr(Type.function()), True)
 
 
+InstructionSource = Optional[Instruction]
+
+
+@dataclass(eq=False)
+class PlannedVar:
+    """A persistent Var, that can be merged with other ones using a
+    union-find-based setup."""
+
+    _parent: Optional["PlannedVar"] = None
+
+    def get_representative(self) -> "PlannedVar":
+        """Find the representative PlannedVar of all the ones that have been
+        merged with this one."""
+        root = self
+        while root._parent is not None:
+            root = root._parent
+        cur = self
+        while cur != root:
+            cur._parent = root
+            cur = cur._parent
```

Comment:
I think the intent was to do something like this? (Alternatively, keep a list of the `root`s visited in the previous while loop.)
```suggestion
        cur = self
        while cur != root:
            next_cur = cur._parent
            cur._parent = root
            cur = next_cur
```